File: sf/models/SfModels.py
from __future__ import annotations

from collections.abc import Mapping, MutableMapping
from enum import Enum
import os
from typing import NamedTuple, TypedDict

CATALOG_NAME: str | None = 'Salesforce'
CATALOG_ALIAS: str | None = 'sf'
CATALOG_NAMESPACE: str | None = 'salesforce'

API_VERSION: str = os.getenv('SF_API_VERSION', '66.0')
SF_EXTERNAL_CLIENT_APP_NAME: str = os.getenv('SF_EXTERNAL_CLIENT_APP_NAME', 'automation')
SF_BASE_URL: str = os.getenv('SF_BASE_URL') or f"https://{os.getenv('SF_BASE_DOMAIN')}.salesforce.com"
SF_CALLBACK_URL: str = os.getenv('SF_CALLBACK_URL', 'http://localhost:1717/OauthRedirect')

# --- Common Types ---
Headers = MutableMapping[str, str]
BulkDataAny = list[Mapping[str, any]]
BulkDataStr = list[Mapping[str, str]]

# --- http methods ---
class HttpMethod(str, Enum):
    delete = 'DELETE'
    get = 'GET'
    head = 'HEAD'
    options = 'OPTIONS'
    patch = 'PATCH'
    post = 'POST'
    put = 'PUT'
    request = 'REQUEST'

# --- REST Models ---
class Usage(NamedTuple):
    """Usage information for a Salesforce org"""
    used: int
    total: int

class PerAppUsage(NamedTuple):
    """Per App Usage information for a Salesforce org"""
    used: int
    total: int
    name: str

# --- Bulk API 2.0 Enums ---
class Operation(str, Enum):
    insert = "insert"
    upsert = "upsert"
    update = "update"
    delete = "delete"
    hard_delete = "hardDelete"
    query = "query"
    query_all = "queryAll"

class JobState(str, Enum):
    open = "Open"
    aborted = "Aborted"
    failed = "Failed"
    upload_complete = "UploadComplete"
    in_progress = "InProgress"
    job_complete = "JobComplete"

class ColumnDelimiter(str, Enum):
    BACKQUOTE = "BACKQUOTE"  # (`)
    CARET = "CARET"          # (^)
    COMMA = "COMMA"          # (,)
    PIPE = "PIPE"            # (|)
    SEMICOLON = "SEMICOLON"  # (;)
    TAB = "TAB"              # (\t)

class LineEnding(str, Enum):
    LF = "LF"
    CRLF = "CRLF"

class ResultsType(str, Enum):
    failed = "failedResults"
    successful = "successfulResults"
    unprocessed = "unprocessedRecords"

# --- Bulk API 2.0 Types ---
class QueryParameters(TypedDict, total=False):
    maxRecords: int
    locator: str

class QueryRecordsResult(TypedDict):
    locator: str
    number_of_records: int
    records: str

QueryResult = QueryRecordsResult

# Objects excluded from migratable describe
SKIP_SUFFIXES = (
    '__History', '__Feed', '__Share', '__Tag',
    '__ChangeEvent', '__e', '__mdt', '__b',
)

SKIP_NAMES = {
    # Feeds
    'AccountFeed', 'ContactFeed', 'CaseFeed', 'LeadFeed',
    'OpportunityFeed', 'UserFeed', 'CollaborationGroupFeed',
    # History
    'AccountHistory', 'ContactHistory', 'CaseHistory',
    'LeadHistory', 'OpportunityHistory', 'OpportunityFieldHistory',
    # Shares
    'AccountShare', 'CaseShare', 'LeadShare', 'OpportunityShare',
    # Apex / Dev
    'ApexClass', 'ApexTrigger', 'ApexLog', 'ApexTestResult',
    'AsyncApexJob', 'CronTrigger', 'CronJobDetail',
    # Content (binary blobs - break bulk migrations)
    'ContentVersion', 'ContentDocument', 'ContentDocumentLink',
    # Restricted query syntax — require specific WHERE filters; can't be queried freely
    'ContentFolderItem', 'IdeaComment',
    # Metadata / Definitions
    'EntityDefinition', 'FieldDefinition', 'FieldPermissions',
    # Auth / Sessions
    'OauthToken', 'AuthSession', 'SessionPermSetActivation',
    'TwoFactorInfo', 'VerificationHistory', 'LoginHistory', 'LoginGeo',
    # Platform
    'StaticResource', 'AuraDefinition', 'AuraDefinitionBundle',
    'FlowDefinitionView', 'FlowInterview',
    'PlatformEventChannel', 'PlatformEventChannelMember',
    'DataStatistics', 'BackgroundOperation', 'SetupAuditTrail',
    # Permissions
    'PermissionSet', 'PermissionSetAssignment',
    'GroupMember', 'UserRole', 'UserLicense',
}

File: typescripts/src/sf/SfModels.ts
/**
 * SfModels.ts
 *
 * Salesforce constants and the request-error type, ported from
 * src/sf/SfModels.py.
 */

export const HttpMethod = {
  delete: 'DELETE',
  get: 'GET',
  head: 'HEAD',
  options: 'OPTIONS',
  patch: 'PATCH',
  post: 'POST',
  put: 'PUT',
  request: 'REQUEST',
} as const
export type HttpMethod = (typeof HttpMethod)[keyof typeof HttpMethod]

export const Operation = {
  insert: 'insert',
  upsert: 'upsert',
  update: 'update',
  delete: 'delete',
  hardDelete: 'hardDelete',
  query: 'query',
  queryAll: 'queryAll',
} as const
export type Operation = (typeof Operation)[keyof typeof Operation]

// Salesforce errorCodes meaning the user lacks rights (skippable per-table
// during a bulk migration rather than fatal).
export const SF_ACCESS_ERROR_CODES: ReadonlySet<string> = new Set([
  'INSUFFICIENT_ACCESS',
  'INSUFFICIENT_ACCESS_OR_READONLY',
  'INSUFFICIENT_ACCESS_ON_CROSS_REFERENCE_ENTITY',
])

// Objects excluded from migratable describe.
export const SKIP_SUFFIXES = [
  '__History',
  '__Feed',
  '__Share',
  '__Tag',
  '__ChangeEvent',
  '__e',
  '__mdt',
  '__b',
] as const

export const SKIP_NAMES: ReadonlySet<string> = new Set([
  // Feeds
  'AccountFeed', 'ContactFeed', 'CaseFeed', 'LeadFeed',
  'OpportunityFeed', 'UserFeed', 'CollaborationGroupFeed',
  // History
  'AccountHistory', 'ContactHistory', 'CaseHistory', 'LeadHistory',
  'OpportunityHistory', 'OpportunityFieldHistory',
  // Shares
  'AccountShare', 'CaseShare', 'LeadShare', 'OpportunityShare',
  // Apex / Dev
  'ApexClass', 'ApexTrigger', 'ApexLog', 'ApexTestResult',
  'AsyncApexJob', 'CronTrigger', 'CronJobDetail',
  // Content (binary blobs - break bulk migrations)
  'ContentVersion', 'ContentDocument', 'ContentDocumentLink',
  // Restricted query syntax
  'ContentFolderItem', 'IdeaComment',
  // Metadata / Definitions
  'EntityDefinition', 'FieldDefinition', 'FieldPermissions',
  // Auth / Sessions
  'OauthToken', 'AuthSession', 'SessionPermSetActivation',
  'TwoFactorInfo', 'VerificationHistory', 'LoginHistory', 'LoginGeo',
  // Platform
  'StaticResource', 'AuraDefinition', 'AuraDefinitionBundle',
  'FlowDefinitionView', 'FlowInterview', 'PlatformEventChannel',
  'PlatformEventChannelMember', 'DataStatistics', 'BackgroundOperation',
  'SetupAuditTrail',
  // Permissions
  'PermissionSet', 'PermissionSetAssignment', 'GroupMember',
  'UserRole', 'UserLicense',
])

/**
 * Raised when the Salesforce REST API returns a non-2xx response. Parses the
 * response body for Salesforce errorCode(s) so callers can react to specific
 * failures (e.g. skipping a table on INSUFFICIENT_ACCESS).
 */
export class SalesforceRequestError extends Error {
  statusCode: number
  method: string
  url: string
  body: string
  errorCodes: string[]

  constructor(statusCode: number, method: unknown, url: unknown, body: string) {
    super(`HTTP ${statusCode} ${String(method)} ${String(url)}: ${body}`)
    this.name = 'SalesforceRequestError'
    this.statusCode = statusCode
    this.method = String(method)
    this.url = String(url)
    this.body = body
    this.errorCodes = SalesforceRequestError._parseErrorCodes(body)
  }

  private static _parseErrorCodes(body: string): string[] {
    let payload: unknown
    try {
      payload = JSON.parse(body)
    } catch {
      return []
    }
    const list = Array.isArray(payload) ? payload : [payload]
    const codes: string[] = []
    for (const item of list) {
      if (item && typeof item === 'object' && 'errorCode' in item) {
        const code = (item as { errorCode?: unknown }).errorCode
        if (code) codes.push(String(code))
      }
    }
    return codes
  }

  /** True when the failure is an access-rights denial (skippable). */
  get isAccessError(): boolean {
    return this.errorCodes.some((code) => SF_ACCESS_ERROR_CODES.has(code))
  }
}
